### PR TITLE
Make the global flag *fDiscover* an instance variable of CConnman

### PR DESCRIPTION
**Pull request by mmachicao**
REFERS TO: #14210

 https://github.com/bitcoin/bitcoin/issues/14210

 - net.h, net.cpp : Removed global flag bool fDiscover
 - net.h, net.cpp : Added field bool fDiscover = true to struct CConnman.Options
 - net.h, net.cpp : Added instance variable bool fDiscover and method bool isDiscover() to CConnman class
 - net.h, net.cpp : Added bool fDiscover to signatures of AddLocal, AdvertiseLocal and IsPeerAddrLocalGood
 - net.h, net.cpp : ThreadMapPort reads fDiscover from the command line, which defaults to *true*

 - test/net_tests.cpp : Test case LocalAddress_BasicLifecycle runs with fDiscovery = true

 - net_processing.cpp : Made PeerLogicValidation use CConnman.isDiscover()

 - torcontrol.cpp : Added instance variable bool fDiscover to TorController class
 - torcontrol.cpp : TorControllerThread reads fDiscover from the command line, which defaults to *true*

 - init.h : AppInitMain reads bool fDiscover from the command line and sets CConnman.Option.fDiscovery

### CODE REVIEW

**Comment by dongcarl** [comment]
Perhaps you want to take a look at [our developer notes](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c) for some naming conventions. Note that the naming convention has changed, and newly added code doesn't need to matching surrounding code. :smiley:

**Comment by MarcoFalke** [comment]
It is not really clear what this is doing and why. Also, it is not clear what this refactoring achieves in the long term.

**Comment by mmachicao** [comment]
> 
> 
> It is not really clear what this is doing and why. Also, it is not clear what this refactoring achieves in the long term.

This PR refers to issue #14210

The issue being that it is difficult to test the network interface thoroughly.

Quotes:

> The idea, again, is to be able to run multiple networking instances inside of one bitcoind process.

> CConnman being instantiable like this would allow us to inspect why a CConnman instance might have stopped receiving data rather than just guessing based on pings and timing checks.

So, I am removing the global dependencies (one by one) until this can be achieved.

**Comment by laanwj** [comment]
Concept ACK, I think if it's possible to have certain variables in a smaller scope, instead of globally external throughout the program, that's preferable.

> Perhaps you want to take a look at our developer notes for some naming conventions.

Yes, please check the naming / coding conventions. We don't use names like `fDiscover` anymore, so this would be a good opportunity to modernize the naming.

File: src/net.cpp [hunk 1]
```diff
@@ -170,29 +169,30 @@ static int GetnScore(const CService& addr)
...
 
 // pushes our own address to a peer
-void AdvertiseLocal(CNode *pnode)
+void AdvertiseLocal(CNode* pnode, bool fDiscover)
 {
     if (fListen && pnode->fSuccessfullyConnected)
     {
         CAddress addrLocal = GetLocalAddress(&pnode->addr, pnode->GetLocalServices());
-        if (gArgs.GetBoolArg("-addrmantest", false)) {
+        if (gArgs.GetBoolArg("-addrmantest", false))
```
Comment (laanwj) [inline src/net.cpp]: Keep this `{` at the end of the line

**Comment by adamjonas** [comment]
Hi @mmachicao - pinging for rebase and squashing the merge conflict commits consistent with the notes found in the [CONTRIBUTING.md doc](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits). Thanks!

**Comment by MarcoFalke** [comment]
Closing for now. Let me know when you want to work on this again

**Pull request closed without merging**